access_request.go

```diff
@@ -7,6 +7,9 @@ type AccessRequest struct {
 	GrantTypes       Arguments `json:"grantTypes" gorethink:"grantTypes"`
 	HandledGrantType Arguments `json:"handledGrantType" gorethink:"handledGrantType"`
 
+	RefreshTokenRequestedScope Arguments
+	RefreshTokenGrantedScope   Arguments
```

Comment:
I assume these two new fields would need to be stored in the database?

Reply:
They shouldn't specifically need to be stored I don't think. The only situation where they're essential is the refresh flow to ensure the value passed to `CreateRefreshTokenStorage` has exactly the same scopes as originally requested/granted.

This will mean the refresh token session will have static scopes in hydra, but the other session types may have narrower scopes than it at any point in the lifetime.

I think this may address some of your other points, and/or may give you a better idea about how we can achieve the desired result. The intention is between `HandleTokenEndpointRequest` and `PopulateTokenEndpointResponse` in the `RefreshTokenGrantHandler`, that we can restore the original scopes of specifically the refresh token session.

The original code is just passing the `Request` struct to the `CreateRefreshTokenStorage` method of `RefreshTokenStorage`, as per the following code section:

```go
	storeReq := requester.Sanitize([]string{})
	storeReq.SetID(ts.GetID())

	if err = c.TokenRevocationStorage.CreateAccessTokenSession(ctx, accessSignature, storeReq); err != nil {
		return err
	}

	if err = c.TokenRevocationStorage.CreateRefreshTokenSession(ctx, refreshSignature, storeReq); err != nil {
		return err
	}
```

That being said, I have just realized over-engineered this a bit. We could just do the following:

```go
	storeReq := requester.Sanitize([]string{})
	storeReq.SetID(ts.GetID())

	if err = c.TokenRevocationStorage.CreateAccessTokenSession(ctx, accessSignature, storeReq); err != nil {
		return err
	}

	if rtRequest, ok := requester.(fosite.RefreshTokenAccessRequester); ok {
		rtStoreReq := requester.Sanitize([]string{}).(*fosite.Request)
		rtStoreReq.SetID(ts.GetID())

		rtStoreReq.RequestedScope = rtRequest.GetRefreshTokenRequestedScopes()
		rtStoreReq.GrantedScope = rtRequest.GetRefreshTokenGrantedScopes()

		if err = c.TokenRevocationStorage.CreateRefreshTokenSession(ctx, refreshSignature, rtStoreReq); err != nil {
			return err
		}
	} else {
		if err = c.TokenRevocationStorage.CreateRefreshTokenSession(ctx, refreshSignature, storeReq); err != nil {
			return err
		}
	}
```

Reply:
Awesome, thank you for the explainer! I think the next step is to create a PR in hydra, update it's go.mod for ory/fosite to cda9c24ce1a981f60d6c2e99b7fceec1654c807d and add an e2e test for this new feature. If that and OIDC conformity pass, I think we're good to go! :)

Reply:
Hey @aeneasr sorry for the delay. This is now done in https://github.com/ory/hydra/pull/3403. Le me know if any adjustments are needed and if you want this test merged. I've fixed the lints locally I believe.

Reply:
Same as above.